pkg/kv/kvserver/replica_write.go

```diff
@@ -884,3 +889,66 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
 	ms.Reset()
 	mvccStatsPool.Put(ms)
 }
+
+// overlapsUnreplicatedRangeIDLocalKeys checks if the provided span overlaps
+// with any unreplicated rangeID local keys.
+// Note that we could receive the span with a nil startKey, which has a special
+// meaning that the span represents: [endKey.Prev(), endKey).
+func overlapsUnreplicatedRangeIDLocalKeys(span spanset.TrickySpan) error {
+	fullRangeIDLocalSpans := roachpb.Span{
+		Key:    keys.LocalRangeIDPrefix.AsRawKey(),
+		EndKey: keys.LocalRangeIDPrefix.AsRawKey().PrefixEnd(),
+	}
+
+	// If the provided span is completely outside the rangeID local spans for any
+	// rangeID, then there is no overlap with any rangeID local keys.
+	if !spanset.Overlaps(fullRangeIDLocalSpans, span) {
+		return nil
+	}
+
+	// At this point, we know that we overlap with fullRangeIDLocalSpans. If we
+	// are not completely within fullRangeIDLocalSpans, return an error as we
+	// collide with at least one unreplicated RangeIDLocal key.
```

Comment:
Technically, it's possible that we don't collide. E.g. you can imagine a little span that begins before the RangeID-local space, and slightly moves into it without touching the first range's unreplicated.

We just make it a "policy" here that it shouldn't be possible, i.e. the spans are sensibly respecting the "tree" structure, and a span that begins outside the subtree and ends inside (or vice versa) doesn't make logical sense.

Dunno it there is concise way to say this, see if you can compress it :)